pkg/reconcilermanager/controllers/reposync_controller.go

```diff
@@ -184,6 +190,90 @@ func (r *RepoSyncReconciler) Reconcile(ctx context.Context, req controllerruntim
 	}
 
+	finalizerName := v1beta1.SyncFinalizer
+
+	// Cross-namespace owner references are disallowed by Kubernetes.
+	// So we can't rely on Kubernetes' built-in garbage collection to delete
+	// managed objects in the central "config-management-system" namespace, when
+	// the RepoSync is in another namespace.
+	// To avoid leaking managed resources, we use a custom finalizer to force
+	// foreground deletion. This way, the RepoSync will not be fully deleted
+	// until all its managed resources are also deleted.
+
+	if rs.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is NOT being deleted.
+		if !controllerutil.ContainsFinalizer(rs, finalizerName) {
+			// The object is new and doesn't have our finalizer yet.
+			// Add our finalizer and update the object.
+			controllerutil.AddFinalizer(rs, finalizerName)
+			err := r.client.Update(ctx, rs)
+			if err != nil {
+				err := status.APIServerError(err, "failed to update RepoSync to add finalizer")
+				log.Error(err, "Finalizer injection failed")
+				metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+				return controllerruntime.Result{}, err
+			}
+			log.Info("Finalizer injection successful")
+		}
+	} else {
+		// The object is being deleted.
+		if controllerutil.ContainsFinalizer(rs, finalizerName) {
+			// Our finalizer is present, so delete managed resource objects.
+			if err := r.deleteManagedObjects(ctx, reconcilerRef, currentRS, rs, true); err != nil {
+				// Failed to delete a managed object.
+				// Return an error to trigger retry.
+				metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+				return controllerruntime.Result{}, errors.Wrap(err, "failed to delete managed objects")
+			}
+			log.Info("Managed object cleanup successful")
+
+			// Remove our finalizer and update the object.
+			controllerutil.RemoveFinalizer(rs, finalizerName)
+			err := r.client.Update(ctx, rs)
+			if err != nil {
+				err := status.APIServerError(err, "failed to update RepoSync to remove finalizer")
+				log.Error(err, "Finalizer removal failed")
+				metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+				return controllerruntime.Result{}, err
+			}
+			log.Info("Finalizer removal successful")
+
+			// cleanup successful
+			r.deleteCachedSync(rsRef)
+		}
+
+		// Stop reconciliation as the item is being deleted
+		metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(nil), start)
+		return controllerruntime.Result{}, nil
+	}
+
+	if err := r.upsertManagedObjects(ctx, reconcilerRef, currentRS, rs); err != nil {
+		metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
+		return controllerruntime.Result{}, err
+	}
+
+	metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(nil), start)
+	return controllerruntime.Result{}, nil
+}
+
+func (r *RepoSyncReconciler) upsertManagedObjects(ctx context.Context, reconcilerRef types.NamespacedName, currentRS, rs *v1beta1.RepoSync) error {
```

Comment:
Extracting upsertManagedObjects could be in a seperate PR, perhaps with the log changes.